amulet/charm.py

```diff
@@ -1,31 +1,64 @@
+from .helpers import reify
+from .helpers import run_bzr
+from .helpers import setup_bzr
+from charmworldlib.charm import Charm
+from path import path
+from path import tempdir
 import os
-import yaml
+import shlex
 import shutil
+import subprocess
 import tempfile
+import yaml
 
-from charmworldlib.charm import Charm
-from .helpers import run_bzr, setup_bzr
 
+class CharmCache(dict):
+    def __init__(self, test_charm):
+        super(CharmCache, self).__init__()
+        self.test_charm = test_charm
+
+    @staticmethod
+    def get_charm(charm_path, branch=None, series='precise'):
+        if charm_path.startswith('lp:'):
+            return LaunchpadCharm(charm_path)
+        elif branch and branch.startswith('lp:'):
+            return LaunchpadCharm(branch)
+
+        if charm_path.startswith('local:'):
+            return LocalCharm(
+                os.path.join(
+                    os.environ.get('JUJU_REPOSITORY', ''),
+                    charm_path[len('local:'):]))
+
+        if branch and branch.endswith('.git'):
+            return GitCharm(branch, name=charm_path)
 
-def get_charm(charm_path, series='precise'):
-    def with_series(charm_path):
-        if '/' not in charm_path:
-            return '{}/{}'.format(series, charm_path)
-        return charm_path
+        if os.path.exists(os.path.expanduser(charm_path)):
+            return LocalCharm(charm_path)
 
-    if charm_path.startswith('cs:'):
-        return Charm(with_series(charm_path))
-    if charm_path.startswith('lp:'):
-        return LaunchpadCharm(charm_path)
-    if charm_path.startswith('local:'):
-        return LocalCharm(
-            os.path.join(
-                os.environ.get('JUJU_REPOSITORY', ''),
-                charm_path[len('local:'):]))
-    if os.path.exists(os.path.expanduser(charm_path)):
-        return LocalCharm(charm_path)
+        return Charm(with_series(charm_path, series))
 
-    return Charm(with_series(charm_path))
+    def __getitem__(self, service):
+        return self.fetch(service)
+
+    def fetch(self, service, charm=None, branch=None, series='precise'):
+        charm_ = charm
+        charm = super(CharmCache, self).get(service, None)
+        if charm is not None:
+            return charm
+
+        charm = charm_ or service
+        charm = os.getcwd() if charm == self.test_charm else charm
+        self[service] = self.get_charm(charm,
+                                       branch=branch,
+                                       series=series)
+        return self.get(service)
+
+
+def with_series(charm_path, series):
+    if '/' not in charm_path:
+        return '{}/{}'.format(series, charm_path)
+    return charm_path
 
 
 def is_branch(path):
```

Comment:
Sorry, it's not clear from this line comment, but this function tests for VCS charms by looking for .bzr directories, and the comment implies it also handles Git, but it doesn't.  It seems like it should also check for the .git directory.